lib/workload/stateless/stacks/fmannotator/handlers.go

```diff
@@ -0,0 +1,45 @@
+// Package fmannotator Lambda handler implementations.
+package fmannotator
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/umccr/orcabus/lib/workload/stateless/stacks/fmannotator/schema/orcabus_workflowmanager/workflowrunstatechange"
+	"log/slog"
+	"net/url"
+	"strings"
+)
+
+// PortalRunId Annotate the portalRunId from an incoming WorkflowRunStateChange event using the config and FM endpoint token.
+func PortalRunId(event workflowrunstatechange.Event, config *Config, token string) (err error) {
+	eventStatus := strings.ToUpper(event.Detail.Status)
+	if eventStatus != "SUCCEEDED" && eventStatus != "FAILED" && eventStatus != "ABORTED" {
+		return nil
+	}
+
+	patch, err := MarshallPortalRunId(&event)
+	if err != nil {
+		return err
+	}
+
+	req, err := NewApiClient(config, bytes.NewBuffer(patch))
+	if err != nil {
+		return err
+	}
+
+	req = req.WithMethod("PATCH").WithS3Endpoint().WithQuery(url.Values{
+		"key": {fmt.Sprintf("*%v*", event.Detail.PortalRunId)},
```

Comment:
Could add the `/` before & after the portalRunId to make sure we are hitting the intended path part and not some random filename / string in some other part of the key.
But I guess the chances of that happening are negligible, especially since we are not using the portalRunId for anything else in the key....

Reply:
Good idea. 

As an additional comment, at some point it might be nice for the `WorkflowRunStateChange` event to also contain an `outputPath` where objects are expected to turn up. This would allow this service to query the filemanager API with `/api/v1/s3?key=<output_path>/<portal_run_id>/*`, which should be faster because it will hit the index that supports prefix matching, and it would probably be even more robust.

Reply:
Hm,... 

To some extend that's already in place... WorkflowRunStateChange events can contain a `Payload` JSON which may contain that information. However, this is not enforced nor standardised. That's on purpose (at least for now) to keep the use cases flexible. We may change that in the future... 

Also, I am not sure that would always work... In the normal case the portalRunId update would happen immediately after the objects have been written so we would not expect a key change. However, especially the prefix of the key can change (e.g. hot storage vs archive) and we try to focus on the key part starting from the portalRunId.
So depending on when this procedure is run, this may or may not work.

Reply:
Okay makes sense. I think we leave it for now and see if it emerges later as a possibility.